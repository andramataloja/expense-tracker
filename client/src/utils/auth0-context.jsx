import React, { useState, useEffect, useContext } from "react";
import createAuth0Client from "@auth0/auth0-spa-js";

const DEFAULT_REDIRECT_CALLBACK = () =>
  window.history.replaceState({}, document.title, window.location.pathname);

export const Auth0Context = React.createContext();
export const useAuth0 = () => useContext(Auth0Context);
export const Auth0Provider = ({
  children,
  onRedirectCallback = DEFAULT_REDIRECT_CALLBACK
}) => {
  const [isAuthenticated, setIsAuthenticated] = useState();
  const [user, setUser] = useState();
  const [auth0Client, setAuth0Client] = useState();
  const [isLoading, setIsLoading] = useState(true);

  const config = {
    domain: process.env.REACT_APP_AUTH0_DOMAIN,
    client_id: process.env.REACT_APP_AUTH0_CLIENT_ID,
    redirect_uri: "http://localhost:3000/Dashboard"
  };

  useEffect(() => {
    const initializeAuth0 = async () => {
      const auth0FromHook = await createAuth0Client(config);
      setAuth0Client(auth0FromHook);

      if (
        window.location.search.includes("code=") &&
        window.location.search.includes("state=")
      ) {
        const { appState } = await auth0FromHook.handleRedirectCallback();
        onRedirectCallback(appState);
      }

      const isAuthenticated = await auth0FromHook.isAuthenticated();

      setIsAuthenticated(isAuthenticated);

      if (isAuthenticated) {
        const user = await auth0FromHook.getUser();
        setUser(user);
      }

      setIsLoading(false);
    };
    initializeAuth0();
  }, []);

  const handleRedirectCallback = async () => {
    setIsLoading(true);
    await auth0Client.handleRedirectCallback();
    const user = await auth0Client.getUser();
    setIsLoading(false);
    setIsAuthenticated(true);
    setUser(user);
  };
  return (
    <Auth0Context.Provider
      value={{
        isAuthenticated,
        user,
        isLoading,
        handleRedirectCallback,
        getIdTokenClaims: (...p) => auth0Client.getIdTokenClaims(...p),
        loginWithRedirect: (...p) => auth0Client.loginWithRedirect(...p),
        getTokenSilently: (...p) => auth0Client.getTokenSilently(...p),
        logout: (...p) => auth0Client.logout(...p)
      }}
    >
      {children}
    </Auth0Context.Provider>
  );
};
