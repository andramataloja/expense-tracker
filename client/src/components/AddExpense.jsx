import React, { useEffect, useState } from "react";
import {
  Button,
  Box,
  TextField,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormControl,
  InputLabel,
  OutlinedInput,
  InputAdornment,
  MenuItem,
  FormGroup,
  Fab,
  makeStyles
} from "@material-ui/core";
import "date-fns";
import DateFnsUtils from "@date-io/date-fns";
import {
  MuiPickersUtilsProvider,
  KeyboardDatePicker
} from "@material-ui/pickers";
import { useAuth0 } from "../utils/auth0-context";
import axios from "axios";
import AddIcon from "@material-ui/icons/Add";

const useStyles = makeStyles(theme => ({
  addButton: {
    color: "white"
  },
  addCommands: {
    color: theme.palette.secondary.dark
  }
}));

const AddExpense = props => {
  const classes = useStyles();
  /* const {user} = useAuth0(); */

  const [open, setOpen] = useState(false);
  const [category, setCategory] = useState("");
  const [selectedDate, setSelectedDate] = useState(new Date());
  const [categoryList, setCategoryList] = useState([]);
  const [values, setValues] = useState({
    amount: "",
    description: ""
  });

  const handleChange = prop => event => {
    setValues({ ...values, [prop]: event.target.value });
    console.log("des", values.description);
    console.log("amount", values.amount);
  };

  const handleCategoryChange = event => {
    setCategory(event.target.value);
    console.log("cat", event.target.value, category);
  };

  const handleDateChange = date => {
    const formattedDate =
      date.getFullYear() + "-" + (date.getMonth() + 1) + "-" + date.getDate();
    setSelectedDate(formattedDate);
    console.log("date", formattedDate);
  };

  const handleSubmit = event => {
    event.preventDefault();
    console.log("des", values.description);
    //const user_id = user.user_id
    const data = {
      description: values.description,
      amount: values.amount,
      date: selectedDate,
      //      user_id: user_id,
      category_id: category
    };

    axios
      .post("/post/expensetodb", data)
      .then(response => console.log("posting", response))
      .catch(err => console.log(err));
    window.location.reload(false);
  };

  useEffect(() => {
    axios
      .get("/get/categories")
      .then(res => (res.data.length !== 0 ? setCategoryList(res.data) : null))
      .catch(err => console.log(err));
  }, []);

  return (
    <Box>
      <Fab color="primary">
        <AddIcon
          fontSize="large"
          onClick={() => setOpen(true)}
          className={classes.addButton}
        />
      </Fab>
      <Dialog
        open={open}
        onClose={() => setOpen(false)}
        aria-labelledby="form-dialog-title"
      >
        <DialogTitle id="form-dialog-title">Add Expense</DialogTitle>
        <DialogContent>
          <FormGroup>
            <FormControl variant="outlined" margin="dense">
              <InputLabel htmlFor="component-outlined">Description</InputLabel>
              <OutlinedInput
                id="component-outlined"
                value={values.description}
                onChange={handleChange("description")}
                label="Description"
              />
            </FormControl>
            <FormControl variant="outlined" margin="dense">
              <InputLabel htmlFor="outlined-adornment-amount">
                Amount
              </InputLabel>
              <OutlinedInput
                id="outlined-adornment-amount"
                value={values.amount}
                onChange={handleChange("amount")}
                startAdornment={
                  <InputAdornment position="start">â‚¬</InputAdornment>
                }
                label="Amount"
                type="number"
                inputProps={{ min: "0" }}
              />
            </FormControl>
            <MuiPickersUtilsProvider utils={DateFnsUtils}>
              <KeyboardDatePicker
                disableToolbar
                variant="inline"
                inputVariant="outlined"
                format="dd/MM/yyyy"
                margin="dense"
                id="date-picker-inline"
                label="Date"
                value={selectedDate}
                onChange={handleDateChange}
                KeyboardButtonProps={{
                  "aria-label": "change date"
                }}
              />
            </MuiPickersUtilsProvider>
            <TextField
              id="outlined-select-category"
              select
              label="Category"
              value={category}
              onChange={handleCategoryChange}
              variant="outlined"
              margin="dense"
            >
              {categoryList.map(option => (
                <MenuItem key={option.category_id} value={option.category_id}>
                  {option.category_name}
                </MenuItem>
              ))}
            </TextField>
          </FormGroup>
        </DialogContent>
        <DialogActions>
          <Button
            onClick={() => setOpen(false)}
            className={classes.addCommands}
          >
            Cancel
          </Button>
          <Button onClick={handleSubmit} className={classes.addCommands}>
            Add
          </Button>
        </DialogActions>
      </Dialog>
    </Box>
  );
};
export default AddExpense;
