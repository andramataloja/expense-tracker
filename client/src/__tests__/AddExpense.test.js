import React from "react";
import {
  cleanup,
  render,
  fireEvent,
  screen,
  wait,
  waitForElement
} from "@testing-library/react";
import AddExpense from "../components/AddExpense";
import { useAuth0 } from "../utils/auth0-context";
import { Provider } from "react-redux";
import { act } from "react-dom/test-utils";

import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import mockAxios from "axios";

const API = "http://localhost:3000";

const user = {
  email: "johndoe@me.com",
  email_verified: true,
  sub: "google-oauth2|12345678901234"
};

jest.mock("../utils/auth0-context");
jest.mock("../store.js");

describe("Add Expense", () => {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);

  afterEach(cleanup);
  beforeEach(() => {
    useAuth0.mockReturnValue({
      isAuthenticated: true,
      user,
      logout: jest.fn(),
      loginWithRedirect: jest.fn()
    });
    require("mutationobserver-shim");
    mockAxios.get.mockImplementationOnce(() => Promise.resolve(mockData));
  });

  const mockData = {
    data: [
      {
        category_id: 3,
        category_name: "Entertainment",
        icon: "Entertainment",
        fill: "#EA6E6E"
      },
      {
        category_id: 1,
        category_name: "Food",
        icon: "Food",
        fill: "#FCD246"
      }
    ]
  };

  it("fetches successfully data from an API", async () => {
    expect(mockAxios.get(`${API}/get/categories`)).resolves.toEqual(mockData);
    expect(mockAxios.get).toHaveBeenCalledWith(`${API}/get/categories`);
  });

  it("should take a snapshot of add button", async () => {
    const store = mockStore();
    const { asFragment } = render(
      <Provider store={store}>
        <AddExpense />
      </Provider>
    );
    await act(wait);

    expect(asFragment(<AddExpense />)).toMatchSnapshot();
  });

  it("add button opens add expense dialog", async () => {
    const store = mockStore();
    jest.useFakeTimers();
    render(
      <Provider store={store}>
        <AddExpense />
      </Provider>
    );
    await act(wait);

    expect(screen.queryByTestId("add-dialog")).toBeNull();

    fireEvent.click(screen.getByTestId("add-button"));
    expect(screen.getByTestId("add-dialog")).toBeVisible();

    fireEvent.click(screen.getByTestId("cancel-button"));
    await wait(() => expect(screen.getByTestId("add-dialog")).toBeNull()); //miks sa ei tööta??
  });

  it("handlechange in dialog", async () => {
    const store = mockStore();
    const handleChange = jest.fn();
    render(
      <Provider store={store}>
        <AddExpense handleChange={handleChange} />
      </Provider>
    );
    await act(wait);

    fireEvent.click(screen.getByTestId("add-button"));

    const input = screen.getByTestId("description").querySelector("input");
    console.log("debug", screen.debug(input));

    fireEvent.change(input, {
      target: { value: "TEST VALUE" }
    });
    expect(input).toHaveValue("TEST VALUE");
    expect(handleChange).toHaveBeenCalledTimes(1); //miks callib 0 korda?
  });
});
