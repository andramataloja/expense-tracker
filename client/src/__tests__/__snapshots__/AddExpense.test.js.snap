// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Expense should take a snapshot of add button 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-5"
  >
    <button
      class="MuiButtonBase-root MuiFab-root MuiFab-primary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root makeStyles-addButton-1 MuiSvgIcon-fontSizeLarge"
          data-testid="add-button"
          focusable="false"
          role="button"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
